Feature 1493 time as byte

hsoh-u opened the pull request:
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

- [x] Recommend testing for the reviewer to perform, including the location of input datasets:</br>

- [x] Will this PR result in changes to the test suite? **No**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] After merging, should the reviewer **DELETE** the feature branch from GitHub? **Yes**</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, and **Milestone**
- [ ] After submitting the PR, select **Linked Issues** with the original issue number.

Updates:
- The API "get_double_var" at nc_utils.cc is renamed to "get_nc_time".
- point2grid: give the proper error message if the input file does not exist
- Added an unit test to unit_netcdf.xml

Note: The location of the test input file:
dakota:/d3/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc
kiowa:/d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc

At dakota:
./plot_data_plane /d3/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc time_as_byte.ps 'name="wind_speed"; level="(1,\*,\*)";' -v 2
OR
At kiowa:
./plot_data_plane /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc time_as_byte.ps 'name="wind_speed"; level="(1,\*,\*)";' -v 2

for point2grid
./point2grid file_does_not_exist.nc G212 dumpy_output.nc -field 'name=\"AOD\"; level=\"\*\";'

hsoh-u committed:
Check if the input file exists to avoid seg fault

hsoh-u committed:
#1493 Renamed get_double_var to get_nc_time and supports 1 byte integer

hsoh-u committed:
#1493 Added netcdf_1byte_time

JohnHalleyGotway committed:
Remove commented-out development code.

JohnHalleyGotway committed:
Removing commented-out development code.

JohnHalleyGotway commented:
Howard and Julie, I took a look at this PR and removed some commented out code from nc_utils.h/.cc. But I realize that I should have checked with Howard first. I removed the unused/undefined get_nc_time() variant. Is that OK?

hsoh-u commented:
Yes, it's OK.

Cheers,
Howard

On Tue, Oct 20, 2020 at 2:36 PM johnhg <notifications@github.com> wrote:

> Howard and Julie, I took a look at this PR and removed some commented out
> code from nc_utils.h/.cc. But I realize that I should have checked with
> Howard first. I removed the unused/undefined get_nc_time() variant. Is that
> OK?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/MET/pull/1529#issuecomment-713124591>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADT7WIR3MOCUZ2VQL5NLKHDSLXYFRANCNFSM4SYRS2KA>
> .
>

jprestop commented on met/src/libcode/vx_nc_util/nc_utils.h:
```diff
@@ -216,8 +216,7 @@ extern char   get_char_val(NcVar *var, const int index);
 extern int    get_int_var(NcFile *, const char * var_name, const int index);
 extern int    get_int_var(NcVar *, const int index);
 
-extern double get_double_var(NcFile *, const char * var_name, const int index = 0);
```
Howard, I noticed that this function disappears entirely and is not renamed.  Is that intended?

hsoh-u replied:
Yes. There were two get_double_var(). There was no call of get_double_var(NcFile *). So I commented out and John deleted it.

Sorry the command is not correct: "level="(1,,)" ==> "level="(1,*,*)"

jprestop reviewed (commented):
Code changes look good.  Just wanting on a reply to my question about a function that is being removed instead of being renamed.  I also still need to test and will do that now.

jprestop commented:
Is this the expected output for the commands you asked me to run?

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/plot_data_plane /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc time_as_byte.ps 'name="wind_speed"; level="(1,,)";' -v 2**
DEBUG 1: Opening data file: /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc
ERROR  : 
ERROR  : get_nc_att(NcFile) -> can't read attribute "TITLE" from "/".
ERROR  : 
ERROR  : 
ERROR  : NcCfFile::getData(NcVar *, const LongArray &, DataPlane &) const -> needed 3 arguments for variable wind_speed, got 1
ERROR  : 

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/point2grid file_does_not_exist.nc G212 dumpy_output.nc -field 'name="AOD"; level="*";'**
ERROR  : 
ERROR  : process_command_line() -> file does not exist "file_does_not_exist.nc"
ERROR  :

hsoh-u commented:
Sorry the commend is not correct. The asterisk character is missing.
level(1,*,*)  ==> level(1,\*,\*)

hsoh-u commented:
Lesson learned: 
The copy/paste does not work with the github issue editor because the asteridsk "\*" is a special character. So I should check the command at the "Preview' tab.

jprestop commented:
Thanks, Howard.  I changed the command and reran.  Is this the expected output for both runs?

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/plot_data_plane /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc time_as_byte.ps 'name="wind_speed"; level="(1,*,*)";' -v 2**
DEBUG 1: Opening data file: /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc
ERROR  : 
ERROR  : get_nc_att(NcFile) -> can't read attribute "TITLE" from "/".
ERROR  : 
DEBUG 1: Creating postscript file: time_as_byte.ps

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/point2grid file_does_not_exist.nc G212 dumpy_output.nc -field 'name="AOD"; level="*";'**
ERROR :
ERROR : process_command_line() -> file does not exist "file_does_not_exist.nc"
ERROR :

hsoh-u commented:
Yes, the expected results except the ERROR log.
I will fix the ERROR log.

jprestop commented:
Ok great.  Please let me know once that is done and I can rerun.  Thanks!

hsoh-u committed:
Update nc_utils.cc

#1493 get_nc_att(NcFile) Do not display the error message if exit_on_error is not enabled

jprestop commented:
Ok.  I reran.  Hopefully this is the expected output for both runs.  One you confirm I'll go ahead and approve and squash and merge.

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/plot_data_plane /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc time_as_byte.ps 'name="wind_speed"; level="(1,*,*)";' -v 2**
DEBUG 1: Opening data file: /d1/projects/MET/MET_test_data/unit_test/model_data/nccf/cyg.ddmi.s20200803-003000-e20200803-233000.l3.grid-wind.a10.d21.nc
DEBUG 1: Creating postscript file: time_as_byte.ps

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> **met/bin/point2grid file_does_not_exist.nc G212 dumpy_output.nc -field 'name="AOD"; level="*";'**
ERROR  : 
ERROR  : process_command_line() -> file does not exist "file_does_not_exist.nc"
ERROR  :

hsoh-u commented:
Yes, these are expected output. I added an unit test to unit_netcdf.xml. Would you check it, too?

test/perl/unit.pl test/xml/unit_netcdf.xml

jprestop commented:
Hrm... I encountered the following problem on kiowa:

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> test/perl/unit.pl test/xml/unit_netcdf.xml
Can't locate VxUtil.pm in @INC (you may need to install the VxUtil module) (@INC contains: /lib /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.24.1 /usr/local/share/perl/5.24.1 /usr/lib/x86_64-linux-gnu/perl5/5.24 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.24 /usr/share/perl/5.24 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at test/perl/unit.pl line 9.
BEGIN failed--compilation aborted at test/perl/unit.pl line 9.

I'm not quite sure how to move forward on that.  In the past, when I have received messages like these, I have asked RAL-IT to install the package, but I'm guessing there is something else that is needed here?

jprestop commented:
I set:
setenv MET_TEST_BASE /d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop/test

setenv MET_BASE /d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop/met

looks like I still need to set up some more vars... I'll keep working on it.

jprestop commented:
Ok.  Got it!  I set the following and it is running now:

setenv MET_TEST_BASE /d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop/test
setenv MET_BASE /d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop/met/share/met
setenv MET_TEST_INPUT /d1/projects/MET/MET_test_data/unit_test
setenv MET_TEST_OUTPUT /d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop/output

jprestop commented:
Good news:

kiowa:jpresto:/d1/projects/MET/MET_pull_requests/met-10.0_beta1/feature_1493/MET-feature_1493_time_as_byte_into_develop> test/perl/unit.pl test/xml/unit_netcdf.xml
TEST: ascii2nc_no_compression           - pass -  14.187 sec
TEST: ascii2nc_compression2_by_config   - pass -  14.257 sec
TEST: ascii2nc_compression3_by_env      - pass -  14.262 sec
TEST: ascii2nc_compression4_by_argument - pass -  14.291 sec
TEST: 365_days                          - pass -   1.633 sec
TEST: netcdf_1byte_time                 - pass -  15.981 sec

I will approve and squash and merge!  Thanks, Howard and John, for all of your help with this.

jprestop reviewed (approved):
Changes have been tested and look good.  I approve.
